x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/custom/CustomRequestManager.java

```diff
@@ -0,0 +1,101 @@
+import org.elasticsearch.xpack.inference.external.http.sender.QueryAndDocsInputs;
+import org.elasticsearch.xpack.inference.services.custom.request.CustomRequest;
+import org.elasticsearch.xpack.inference.services.custom.response.CustomResponseEntity;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+public class CustomRequestManager extends BaseRequestManager {
+    private static final Logger logger = LogManager.getLogger(CustomRequestManager.class);
+
+    record RateLimitGrouping(int apiKeyHash) {
+        public static RateLimitGrouping of(CustomModel model) {
+            Objects.requireNonNull(model);
+
+            return new RateLimitGrouping(model.rateLimitServiceSettings().hashCode());
+        }
+    }
+
+    private static ResponseHandler createCustomHandler(CustomModel model) {
+        return new CustomResponseHandler("custom model", CustomResponseEntity::fromResponse, model.getServiceSettings().getErrorParser());
+    }
+
+    public static CustomRequestManager of(CustomModel model, ThreadPool threadPool) {
+        return new CustomRequestManager(Objects.requireNonNull(model), Objects.requireNonNull(threadPool));
+    }
+
+    private final CustomModel model;
+    private final ResponseHandler handler;
+
+    private CustomRequestManager(CustomModel model, ThreadPool threadPool) {
+        super(threadPool, model.getInferenceEntityId(), RateLimitGrouping.of(model), model.rateLimitServiceSettings().rateLimitSettings());
+        this.model = model;
+        this.handler = createCustomHandler(model);
+    }
+
+    @Override
+    public void execute(
+        InferenceInputs inferenceInputs,
+        RequestSender requestSender,
+        Supplier<Boolean> hasRequestCompletedFunction,
+        ActionListener<InferenceServiceResults> listener
+    ) {
+        String query;
+        List<String> input;
+        if (inferenceInputs instanceof QueryAndDocsInputs) {
+            QueryAndDocsInputs queryAndDocsInputs = QueryAndDocsInputs.of(inferenceInputs);
+            query = queryAndDocsInputs.getQuery();
+            input = queryAndDocsInputs.getChunks();
+        } else if (inferenceInputs instanceof ChatCompletionInput chatInputs) {
+            query = null;
+            input = chatInputs.getInputs();
+        } else if (inferenceInputs instanceof EmbeddingsInput) {
+            EmbeddingsInput embeddingsInput = EmbeddingsInput.of(inferenceInputs);
+            query = null;
+            input = embeddingsInput.getStringInputs();
+        } else {
+            listener.onFailure(
+                new ElasticsearchStatusException(
+                    Strings.format("Invalid input received from custom service %s", inferenceInputs.getClass().getSimpleName()),
+                    RestStatus.BAD_REQUEST
+                )
+            );
+            return;
+        }
+
+        try {
+            var request = new CustomRequest(query, input, model);
```

Comment:
`EmbeddingsInput` has a `inputType` parameter that should be passed to the `CustomRequest` so that is can be replaced in the request body. Same for the `topN` and `returnDocuments` options  in `QueryAndDocsInputs`, these could all be passed to the `CustomRequest` constructor as a loose map